implementations/rust/vault/src/error.rs

```diff
@@ -0,0 +1,281 @@
+    #[fail(display = "Failed to find the specified public key")]
+    PublicKey,
+    /// Failed to compute elliptic curve diffie-hellman
+    #[fail(display = "Failed to compute elliptic curve diffie-hellman")]
+    Ecdh,
+    /// Failed to compute HKDF SHA-256 digest
+    #[fail(display = "Failed to compute HKDF SHA-256 digest")]
+    HkdfSha256,
+    /// Failed to encrypt data with AES-GCM
+    #[fail(display = "Failed to encrypt data with AES-GCM")]
+    AeadAesGcmEncrypt,
+    /// Failed to decrypt data with AES-GCM
+    #[fail(display = "Failed to decrypt data with AES-GCM")]
+    AeadAesGcmDecrypt,
+    /// Could not use the AES-GCM cipher scheme
+    #[fail(display = "Could not use the AES-GCM cipher scheme")]
+    AeadAesGcm,
+    /// An invalid parameter was supplied: {}
+    #[fail(display = "An invalid parameter was supplied: {}", 0)]
+    InvalidParam(usize),
+    /// Invalid attributes were specified
+    #[fail(display = "Invalid attributes were specified")]
+    InvalidAttributes,
+    /// An invalid context was supplied
+    #[fail(display = "An invalid context was supplied")]
+    InvalidContext,
+    /// An invalid buffer was supplied
+    #[fail(display = "An invalid buffer was supplied")]
+    InvalidBuffer,
+    /// An invalid size was supplied
+    #[fail(display = "An invalid size was supplied")]
+    InvalidSize,
+    /// An invalid key regeneration occurred
+    #[fail(display = "An invalid key regeneration occurred")]
+    InvalidRegenerate,
+    /// An invalid secret was supplied
+    #[fail(display = "An invalid secret was supplied")]
+    InvalidSecret,
+    /// Invalid secret attributes were supplied
+    #[fail(display = "Invalid secret attributes were supplied")]
+    InvalidSecretAttributes,
+    /// An invalid secret type was supplied that is not supported
+    #[fail(display = "An invalid secret type was supplied that is not supported")]
+    InvalidSecretType,
+    /// An invalid tag was supplied for decryption
+    #[fail(display = "An invalid tag was supplied for decryption")]
+    InvalidTag,
+    /// The supplied buffer was too small
+    #[fail(display = "The supplied buffer was too small")]
+    BufferTooSmall,
+    /// Default requires a specified random generator
+    #[fail(display = "Default requires a specified random generator")]
+    DefaultRandomRequired,
+    /// Default requires a specified memory handler
+    #[fail(display = "Default requires a specified memory handler")]
+    MemoryRequired,
+    /// The secret size specified does not match the expected value
+    #[fail(display = "The secret size specified does not match the expected value")]
+    SecretSizeMismatch,
+    /// An error occurred while reading from I/O
+    #[fail(display = "An error occurred while reading from I/O")]
+    IOError,
+}
+
+impl VaultFailErrorKind {
+    pub(crate) const ERROR_INTERFACE_VAULT: usize = 3 << 24;
+    /// Convert to an integer
+    pub fn to_usize(&self) -> usize {
```

Comment:
@mrinalwadhwa @robinbudd there was an idea of revisiting error handling in C, which probably will also affect this part of a rust code. Any news on that?

Reply:
Yes, but let's not block the rust code on that change.
We can iterate one both pieces together, once they're working together.